src/python/gem5/utils/multisim/multisim.py

```diff
@@ -0,0 +1,299 @@
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""This module contains the gem5 MultiSim framework. The gem5 MultiSim
+work functions by allowing a user to specify multiple simulations to run from
+a single gem5 config script. The MuliSim framework will then run these
+simulations in parallel using the Python multiprocessing library.
+
+The framework works by having the user add the simulators in a configuration
+via the `add_simulator`function, each with a unique, user specified, id. This
+adds the different simulations to run in parallel to a global list. The
+MultiSim framework then uses the Python multiprocessing library to run the
+simulations in parallel by loading the config script as a module in each child
+process and then selecting the simulation to run via the iid in the global set
+of scheduled simulators jobs to run.
+The only difference between the child processes is the id of the simulator.
+
+Important notes
+---------------
+
+1. You cannot load/instnatiate the simulators in the main process. You cannot
+even load the config script (i.e., `import config_script`). This means the
+config script is passed as a string referencing the config script as a modeule.
+This script is then passed to the child processes to load.
+
+2. The child processes run `simulator.run()`. Ergo, users can't specify the
+parameters passed to the `run` function (e.g., `run(max_ticks=1000)`). However,
+in such cases the `Simulator` module should be refactored to have these
+parameters set prior to calling `run`.
+
+3. The config script cannot accept parameters. It must be parameterless.
+"""
+
+import importlib
+import multiprocessing
+from pathlib import Path
+from typing import (
+    Optional,
+    Set,
+)
+
+# A global variable which __main__.py flips to `True` when multisim is run as
+# an executable module.
+module_run = False
+
+
+class SimulatorJob:
+    def __init__(self, simulator: "Simulator", id: str):
```

Comment:
I think it would be good to have a default for the ids, maybe somthing like `workload-id-version`?

I think the easiest way to do this would be to have a `get_id` for `Simulater` class which would default value to id and version of the resource it is suppossed to run (we can get those as all resource clas have a `get_id` and `get_version` functions). We can set the default id to be the `SimulaterJob` id.

while typing this out, I also realised that this would not be as straight forward if we are not using workloads as:

- What if we are using local resources, generally we users wont give and id and version for these.
- What if we are using something the is not `set_workload`, but lets say `set_kernel_disk_worklaod` then how would this work?

This is just a small thing that I feel would be convinent, it is not necessay.

Reply:
I've kind of went with this. I've added IDs to Simulators but left it up to the multisim module to add a default if none were provided. I spent too long experimenting with a "hash as ID" idea but it just provided too complex to work.